src/vsr/README.md

```diff
@@ -0,0 +1,187 @@
+
+## Protocol: Ping (Replica-Replica)
+
+Replicas send `command=ping`/`command=pong` messages to one another to synchronize clocks.
+
+Additionally, both of these messages serve as heartbeats to let replicas know which _other_ replicas are healthy, i.e. reachable and in `status=normal`.
+If a [replication quorum](#quorums) of replicas is not healthy, begin [Start-View-Change protocol](#protocol-start-view-change).
+
+## Protocol: Ping (Replica-Client)
+
+Clients send `command=ping_client` (and receive `command=pong_client`) messages to (from) replicas to learn the cluster's current view.
+
+## Protocol: Normal
+
+Normal protocol prepares and commits requests (from clients) and sends replies (to clients).
+
+1. The client sends a `command=request` message to the primary. (If the client's view is outdated, the receiver will forward the message on to the actual primary).
+2. The primary converts the `command=request` to a `command=prepare` (assigning it an `op` and `timestamp`).
+3. Each replica (in a chain beginning with the primary) performs the following steps concurrently:
+    - Write the prepare to the WAL.
+    - Forward the prepare to the next replica in the chain.
+4. Each replica sends a `command=prepare_ok` message to the primary once it has written the prepare to the WAL.
+5. When a primary collects a [replication quorum](#quorums) of `prepare_ok`s _and_ it has committed all preceding prepares, it commits the prepare.
+6. The primary replies to the client.
+7. The backups are informed that the prepare was committed by either:
+    - a subsequent prepare, or
+    - a periodic `command=commit` message.
+
+```mermaid
+sequenceDiagram
+    participant C0 as Client
+    participant R0 as Replica 0 (primary)
+    participant R1 as Replica 1 (backup)
+    participant R2 as Replica 2 (backup)
+
+    C0->>R0: Request A
+
+    R0->>+R0: Prepare A
+    R0->>+R1: Prepare A
+    R1->>+R2: Prepare A
+
+    R0->>-R0: Prepare-Ok A
+    R1->>-R0: Prepare-Ok A
+    R0->>C0: Reply A
+    R2->>-R0: Prepare-Ok A
+```
+
+See also:
+
+  - [VRR](https://pmg.csail.mit.edu/papers/vr-revisited.pdf) §4.1
+
+## Protocol: Start-View-Change
+
+Start-View-Change (SVC) protocol initiates [view-changes](#protocol-view-change) with minimal disruption.
+
+Unlike the Start-View-Change described in [VRR](https://pmg.csail.mit.edu/papers/vr-revisited.pdf) §4.2, this protocol runs in both `status=normal` and `status=view_change` (not just `status=view_change`).
+
+1. Depending on the replica's status:
+    - `status=normal` + primary: If the replica has not received a `prepare_ok` recently (and it expects to), send a `command=start_view_change` to all replicas (including self).
+    - `status=normal` + primary/backup: If the replica has not received a quorum of [heartbeats](#protocol-ping-replica-replica), send a `command=start_view_change` to all replicas (including self).
```

Comment:
Should we say "quorum, which includes the primary" here? As written, I think this'll break in the case where the primary is symmetrically partitioned from cluster.

Thinking more about this, maybe we don't even need a quorum here? Would the following view change condition be enough?:

* on primary: we didn't get enough `prepare_ok`s
* on backup: we didn't get a ping from primary

Reply:
Ah, that's a really good point! And if heartbeats are always _primary → backup_, we could use `commit` messages as heartbeats instead of `ping`/`pong`. Which means `ping`/`pong` can go back to not needing a `view` number.

Reply:
Or maybe use the head `op` number (since it is monotonically increasing). That might come in handy for implementing state transfer, to decide whether we are too far behind to repair the WAL.

Reply:
> And we probably can do something like including cluster time in the no-op commit to make it easy to detect stale ones.
Nevermind, `op` probably wouldn't work for this purpose since it wouldn't increase when the cluster is idle+healthy.